bindings/c/test/mako/stats.hpp

```diff
@@ -0,0 +1,157 @@
+/*
+ * stats.hpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MAKO_STATS_HPP
+#define MAKO_STATS_HPP
+
+#include <array>
+#include <cstdint>
+#include <cstring>
+#include <list>
+#include <utility>
+#include "operations.hpp"
+#include "time.hpp"
+
+namespace mako {
+
+/* size of each block to get detailed latency for each operation */
+constexpr const size_t LAT_BLOCK_SIZE = 4095;
+
+/* memory block allocated to each operation when collecting detailed latency */
+class LatencySampleBlock {
+	uint64_t samples[LAT_BLOCK_SIZE]{
+		0,
+	};
+	uint32_t index{ 0 };
+
+public:
+	LatencySampleBlock() noexcept = default;
+	bool full() const noexcept { return index >= LAT_BLOCK_SIZE; }
+	void put(timediff_t td) {
+		assert(!full());
+		samples[index++] = toIntegerMicroseconds(td);
+	}
+	// return {data block, number of samples}
+	std::pair<uint64_t const*, size_t> data() const noexcept { return { samples, index }; }
+};
+
+/* collect sampled latencies */
+class LatencySampleBin {
+	std::list<LatencySampleBlock> blocks;
+
+public:
+	void reserveOneBlock() {
+		if (blocks.empty())
+			blocks.emplace_back();
+	}
+
+	void put(timediff_t td) {
+		if (blocks.empty() || blocks.back().full())
```

Comment:
We should check OOM situation or cap the size of the bin, and stop collecting stats. One may want to run mako for an extensive period of time.

Reply:
Good point!